example-plugins/sethome.lua

```diff
@@ -0,0 +1,94 @@
+local lukkit = require("lukkit")
+
+local homestore_opts = {
+  store_type: lukkit.store.type.JSON,
+  overwrite_oncorrupt: false
+}
+
+
+-- Runs when the plugin is enabled
+function lukkit.plugin:enable()
+  local plugin_name = self.info:get_name();
+  self.logger:info("Enabling {}", plugin_name)
+end
+
+-- Runs when the plugin is disabled (shutdown, reload etc.)
+function lukkit.plugin:disable()
+  local plugin_name = self.info:get_name();
+  self.logger:info("Disabling {}", plugin_name)
+end
+
+-- Runs when the plugin is loaded (before enabled, on file read, can stop loading)
+function lukkit.plugin:load(load_context)
+  -- Can load a SQLite DB, JSON doc or YAML
+  local home_storage = self.store:load("home", homestore_opts)
+
+  -- do setup with store
+
+  if (home_storage:is_read_only()) then
+    -- Stop the plugin from continuining with a load
+    load_context:reject("Home store cannot be written to")
+  end
+end
+
+
+function lukkit.plugin:command(cmd, sender, args)
+  -- Handles any case permutation of "home" (e.g. "home", "Home", "HOME" etc.)
+  if (cmd:equals_lower("home")) then
+
+    -- Functional wrapper to only execute if the sender is a player, no-op if not
+    sender:if_player(function(player)
+        -- Gets the player's current display name. May or may not be the account name
+        local name = player:display_name()
+        -- Fetch the home store. If not found the resulting LOption will not have a value
+        local home_store = self.store:get("homes")
+
+        -- Will throw an exception & fail the current command (safely)
+        -- When in dev mode debug data will also be shown
+        -- This should only be used when something SHOULD NEVER be false (such as a
+        --   store which was loaded in the 'load' lifecycle stage)
+        lukkit.assert(home_store:has_value(), "No home store was created before read")
+
+        -- Gets the player's UUID which is the home value's key
+        local home_key = player:uuid()
+
+        -- Do a store lookup using the UUID as a key
+        local player_home = home_store:get(home_key)
+
+        player_home:if_exists(function(value)
```

Comment:
DISC: optionals

Reply:
if `player_home` would be an Option type, then this would be straightforward. Would it make sense to use existing terminology? Kotlin uses `let` to handle side-effects on Optional type, some specs (like `fantasy-land`) allow unwrapping using `getOrElse`

Reply:
`player_home` is an Option type. I think my usage of it here is the issue here. Not sure I understood your point.

Reply:
I guess I am not sure how FP looks in Lua. Are you planning to use some popular library for FP in Lua with Option type implementation? Or is this going to be written in Java and translated to Lua with wrappers? Or custom implementation in Lua? I guess I am confused with where the standard lib will live in the project (Lua side or Java side)

Reply:
I don't plan on using a Java/Kot lib with a wrapper for Lua - I'd rather have a custom lib in Java that focuses on being Lua-like. Impl would end up in the shared layer of Lukkit - the same place v2's "wrappers" sat.

Reply:
I have conflicting thoughts about this, but for now I'm for it, we just need to implement this in a way that it's easy to understand. I think we really need to do some "market testing"